crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs

```diff
@@ -0,0 +1,112 @@
+use crate::ConnectUrl;
+use crate::Port;
+use serde::Deserialize;
+use serde::Serialize;
+use std::num::ParseIntError;
+use url::Host;
+
+/// A combination of a host and a port number.
+///
+/// This type is used for serializing and deserializing host strings with a port
+/// optionally present, like `my-tenant.cumulocity.com` or
+/// `mqtt.my-tenant.cumulocity.com:2137`. The port can be omitted in the parsed
+/// string, because user code can easily derive a reasonable default port number
+/// from the context this type is used in, e.g. 443 for HTTPS or 8883 for MQTT
+/// TLS. To easily specify this fallback port, a const type parameter is used.
+///
+/// # Examples
+///
+/// ```
+/// # use tedge_config::{HostPort, Port};
+/// // use a fallback port if not present in string
+/// let http = HostPort::<443>::try_from("my-tenant.cumulocity.com".to_string()).unwrap();
+/// assert_eq!(http.port(), Port(443));
+///
+/// // allow port to be overridden using standard `:PORT` notation
+/// let http = HostPort::<443>::try_from("my-tenant.cumulocity.com:8080".to_string()).unwrap();
+/// assert_eq!(http.port(), Port(8080));
+///
+/// // return error for malformed host strings
+/// assert!(HostPort::<443>::try_from("my-tenant.cumulocity.com:8080:443".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from(":8080".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from("[:::1]:8080".to_string()).is_err());
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(into = "String", try_from = "String")]
+pub struct HostPort<const P: u16> {
```

Comment:
I'm okay to define the default port using a `const` generic.

- First, because using a `Service` trait and phantom values to attach a default port to a struct seems a bit too involved for little gains. I'm not sure how this will work will the derive traits (Debug, Clone, PartialEq, Eq, Serialize, Deserialize).
- Second, because we can improve clarity by adding constants as `const MQTTS : u16 = 8883` and get nicer types as `mqtt_host: HostPort<MQTT>`. 

About the type name.
- I agree that `HostPort` sounds a bit heavy, but at least this conveys something.
- You propose `ServiceHost` as an alternative, I would prefer `ServiceAddress`.
- I let you make a choice.

Reply:
> - You propose `ServiceHost` as an alternative, I would prefer `ServiceAddress`.

`ServiceAddress` doesn't sit well with me, because I think DNS names are not addresses in and of themselves, we look them up to eventually resolve them to addresses.

So while maybe not the most graceful name, we can stay with `HostPort`. It can always be changed and the documentation explains its workings to a sufficient degree.